### ISSUE TITLE
Uncatchable error thrown by exiv2

### ISSUE THREAD

**Issue by WNiels** [LeoHsiao1/pyexiv2#8]
**Bug description:**
While reading some of my images, i get the following error printed to console:

> Error: Upper boundary of data for directory Casio2, entry 0x2000 is out of bounds: Offset = 0x00000822, size = 31578, exceeds buffer size by 76 Bytes; truncating the entry

This error seems to be thrown by exiv2 but is not translated into a catchable python error.

**Expected behaviour:**
The error should be catchable.
A UpperBoundaryError extending Exception should be thrown.
The following code should be able to catch this error. 
```
try:
    exif = im.read_exif() # where img is a valid pyexiv2.Image object
except BaseException as e:
    print(e)
```
**What i found so far:**
I traced the error down to the following function in `core.py`:
```
    def _open_image(self):
        """ Let C++ program open an image and read its metadata,
        save as a global variable in C++ program. """
        api.open_image.restype = ctypes.c_char_p
        ret = api.open_image(
            self.filename).decode()  # WIP: "Error: Upper boundary of data (...)" seems to originate here.
        if ret != OK:
            raise RuntimeError(ret)
```
**Edit**
Added new findings.

**Comment by LeoHsiao1** [comment]
Hi!
I've noticed that some of the exceptions thrown by exiv2 are not caught properly by python, and the exception you mentioned is one of them, which is defined [here].(https://github.com/Exiv2/exiv2/blob/1125dcb1e424dab1c9884d4dccdada0385511bb1/src/tiffvisitor_int.cpp#L1614)
I will try to deal with this problem next week.
Please send me a picture that can raise this exception so that I can debug it.

**Comment by WNiels** [comment]
Thanks for the fast reply, i have sent you an email with a picture for debugging.

**Comment by LeoHsiao1** [comment]
Hi!
I've been a bit lazy lately, so I just released the new version today.
In version 2.0.0, stderr's log 'EXV ERROR' will be converted to an exception. For example:
```python
>>> from pyexiv2 import Image
>>> img = Image(r'c:\Users\Leo\Desktop\CIMG7607_recompress.JPG') 
Traceback (most recent call last):   
  File "<stdin>", line 1, in <module>
  File "C:\Users\Leo\AppData\Local\Programs\Python\Python38\lib\site-packages\pyexiv2\core.py", line 19, in __init__
    self.img = api.open_image(filename.encode(encoding))
RuntimeError: Upper boundary of data for directory Casio2, entry 0x2000 is out of bounds: Offset = 0x00000822, size = 30384, exceeds buffer size by 76 Bytes; truncating the entry
```


In addition, I have solved the UnicodeDecodeError you mentioned in the email, by adding a parameter `encoding='utf-8'` to some methods. For example:
```python
>>> img.read_exif(encoding='ISO-8859-1')
{'Exif.Image.Make': 'CASIO COMPUTER CO.,LTD ', 'Exif.Image.Model': 'EX-Z50 ', ...}
```

**Comment by LeoHsiao1** [comment]
Oops! I found that in the first case, your image will not be opened.
The error log of exiv2 will not interrupt the program, but it will become fatal after it is converted into an exception.
So what about making it an optional feature?
For example:
```python
>>> import pyexiv2
>>> img = pyexiv2.Image(r'c:\Users\Leo\Desktop\CIMG7607_recompress.JPG')
Error: Upper boundary of data for directory Casio2, entry 0x2000 is out of bounds: Offset = 0x00000822, size = 30384, exceeds buffer size by 76 Bytes; truncating the entry
>>> pyexiv2.convert_error_log_to_exception()    # Enable
>>> img = Image(r'c:\Users\Leo\Desktop\CIMG7607_recompress.JPG')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\1\pyexiv2\pyexiv2\core.py", line 21, in __init__
    self.img = api.open_image(filename.encode(encoding))
RuntimeError: Upper boundary of data for directory Casio2, entry 0x2000 is out of bounds: Offset = 0x00000822, size = 30384, exceeds buffer size by 76 Bytes; truncating the entry
```

**Comment by WNiels** [comment]
Hi,

thanks so far! I have also been busy.But I will try to test your changes
this week and give you feedback!

With kind regards,

*Niels Westphal*
mail@nielswestphal.de
Tel.: 0176 66874143

**Comment by LeoHsiao1** [comment]
I added a function [pyexiv2.set_log_level()](https://github.com/LeoHsiao1/pyexiv2/blob/ea1d4b19c25783068e64b39869e09de122c4da4d/docs/Tutorial.md#logs) to [the dev branch](https://github.com/LeoHsiao1/pyexiv2/tree/ea1d4b19c25783068e64b39869e09de122c4da4d).

You can use the following code:
```python
>>> import pyexiv2
>>> img = pyexiv2.Image(r'c:\Users\Leo\Desktop\CIMG7607_recompress.JPG')
RuntimeError: Upper boundary of data for directory Casio2, entry 0x2000 is out of bounds: Offset = 0x00000822, size = 30384, exceeds buffer size by 76 Bytes; truncating the entry
>>> pyexiv2.set_log_level(4)      # Ignore the error log
>>> img = pyexiv2.Image(r'c:\Users\Leo\Desktop\CIMG7607_recompress.JPG')

>>> data = img.read_exif()
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
>>> data = img.read_exif(encoding='ISO-8859-1')      # Use an appropriate encoding
>>> img.close()
```

If there are no problems, I will add this feature to version `2.1.0`.

**Comment by LeoHsiao1** [comment]
I have released version [2.1.0](https://github.com/LeoHsiao1/pyexiv2/releases/tag/v2.1.0)
